staging/src/k8s.io/apiserver/plugin/pkg/audit/webhook/webhook.go

```diff
@@ -0,0 +1,252 @@
+	if err != nil {
+		return nil, err
+	}
+
+	b := &AuditBackend{w: w}
+
+	switch mode {
+	case ModeBatch:
+		// enable batching.
+		b.buffer = make(chan *auditinternal.Event, defaultBatchBufferSize)
+		b.maxBatchSize = defaultBatchMaxSize
+		b.maxBatchWait = defaultBatchMaxWait
+	case ModeBlocking:
+	default:
+		return nil, fmt.Errorf("webhook mode %q is not in list of known modes (%s)",
+			mode, strings.Join(AllowedModes, ","))
+	}
+
+	return b, nil
+}
+
+// Run starts the webhook and shuts down once stopCh closes. It MUST be called before
+// using the backend. Run is non-blocking.
+func (a *AuditBackend) Run(stopCh <-chan struct{}) error {
+	if a.buffer == nil {
+		// Not buffered, nothing to do.
+		return nil
+	}
+
+	f := func() {
+		// Recover from any panics caused by this method so a panic in the
+		// goroutine can't bring down the main routine.
+		defer runtime.HandleCrash()
+
+		t := time.NewTicker(a.maxBatchWait)
+		defer t.Stop() // Release ticker resources
+
+		// Ignore the error result.
+		a.sendBatchEvents(stopCh, t.C)
+	}
+
+	go func() {
+		for {
+			f()
+		}
+	}()
+
+	return nil
+}
+
+// ProcessEvents attempts to POST a serialized audit event to an external service. The actual
+// payload contains an audit.EventList with a single element. The audit webhook may batch
+// multiple events in the future.
+func (a *AuditBackend) ProcessEvents(ev ...*auditinternal.Event) {
+	if a.buffer != nil {
+		a.queueEvents(ev...)
+		return
+	}
+
+	if err := a.sendEvents(ev...); err != nil {
+		glog.Errorf("failed to POST webhook events: %v", err)
+	}
+}
+
+var cloner = conversion.NewCloner()
+
+// queueEvents adds the provided events to the internal buffer.
+func (a *AuditBackend) queueEvents(ev ...*auditinternal.Event) {
+	for _, e := range ev {
+		// Per the audit.Backend interface these events are reused after being
+		// sent to the Sink. Deep copy and send the copy to the queue.
+		event := new(auditinternal.Event)
+		if err := auditinternal.DeepCopy_audit_Event(e, event, cloner); err != nil {
+			glog.Errorf("failed to clone audit event: %v: %#v", err, e)
+			return
+		}
+
+		select {
+		case a.buffer <- event:
+		default:
+			glog.Errorf("events queue blocked, failed to send event %#v", e)
+		}
+	}
+}
+
+// sendBatchEvents attempts to batch some number of events to the backend. It POSTs events
+// in a goroutine and returns a channel that can be used to wait on any error encountered
+// during the POST.
+//
+// The following things can cause sendBatchEvents to exit:
+//
+//   * Some maximum number of events are received.
+//   * Timer has passed, all queued events are sent.
+//   * StopCh is closed, all queued events are sent.
+//
+func (a *AuditBackend) sendBatchEvents(stopCh <-chan struct{}, timer <-chan time.Time) <-chan error {
+	var (
+		events []auditinternal.Event
+		// errCh is used during testing so tests can wait on the error from the
+		// webhook POST. It is not waited on outside of tests.
+		errCh = make(chan error, 1)
```

Comment:
I don't like this errCh that is only used for testing. I think it's confusing since nothing is listening on it. I'd prefer to simply not check for the error in the test. As long as the server receives an accurate audit event, that should be sufficient.